worker/src/sidechain_storage/storage.rs

```diff
@@ -0,0 +1,1033 @@
+	/// database
+	db: SidechainDB,
+	/// shards in database
+	shards: Vec<ShardIdentifierFor<SignedBlock>>,
+	/// map to last sidechain block of every shard
+	last_blocks: HashMap<ShardIdentifierFor<SignedBlock>, LastSidechainBlock>,
+}
+
+impl<SignedBlock: SignedBlockT> SidechainStorage<SignedBlock> {
+	/// loads the DB from the given paths and stores the listed shard
+	/// and their last blocks in memory for better performance
+	pub fn new(path: PathBuf) -> Result<SidechainStorage<SignedBlock>> {
+		// load db
+		let db = SidechainDB::open_default(path)?;
+		let mut storage = SidechainStorage { db, shards: vec![], last_blocks: HashMap::new() };
+		storage.shards = storage.load_shards_from_db()?;
+		// get last block of each shard
+		for shard in storage.shards.iter() {
+			if let Some(last_block) = storage.load_last_block_from_db(shard)? {
+				storage.last_blocks.insert(*shard, last_block);
+			} else {
+				// an empty shard sidechain storage should not exist. Consider deleting this shard from the shards list.
+				error!("Sidechain storage of shard {:?} is empty", shard);
+			}
+		}
+		Ok(storage)
+	}
+
+	/// gets all shards of currently loaded sidechain db
+	pub fn shards(&self) -> &Vec<ShardIdentifierFor<SignedBlock>> {
+		&self.shards
+	}
+
+	/// gets the last block of the current sidechain DB and the given shard
+	pub fn last_block_of_shard(
+		&self,
+		shard: &ShardIdentifierFor<SignedBlock>,
+	) -> Option<&LastSidechainBlock> {
+		self.last_blocks.get(shard)
+	}
+
+	/// gets the block hash of the sidechain block of the given shard and block number, if there is such a block
+	pub fn get_block_hash(
+		&self,
+		shard: &ShardIdentifierFor<SignedBlock>,
+		block_number: BlockNumber,
+	) -> Result<Option<BlockHash>> {
+		self.db.get((*shard, block_number))
+	}
+
+	/// gets the block of the given blockhash, if there is such a block
+	#[allow(unused)]
+	pub fn get_block(&self, block_hash: &BlockHash) -> Result<Option<SignedBlock>> {
+		self.db.get(block_hash)
+	}
+
+	/// update sidechain storage
+	pub fn store_blocks(&mut self, blocks_to_store: Vec<SignedBlock>) -> Result<()> {
+		let mut batch = WriteBatch::default();
+		let mut new_shard = false;
+		for signed_block in blocks_to_store.into_iter() {
+			// check if current block is the next in line
+			let current_shard = &signed_block.block().shard_id();
+			let current_block_nr = signed_block.block().block_number();
+			if self.shards.contains(current_shard) {
+				if let Some(last_block) = self.last_block_of_shard(current_shard) {
+					if last_block.number != current_block_nr - 1 {
+						error!("[Sidechain DB] Sidechainblock (nr: {:?}) is not a succession of the previous block (nr: {:?}) in shard: {:?}",
+						current_block_nr, last_block.number, *current_shard);
+						continue
+					}
+				} else {
+					error!(
+						"[Sidechain DB] Shard {:?} does not have a last block. Skipping block (nr: {:?}) inclusion",
+						*current_shard, current_block_nr
+					);
+					continue
+				}
```

Comment:
This part I had to read three times until I understood it. Maybe extract it to a descriptive function like: `verify_block_ancestry`?

Reply:
Added, thanks for the input.